kingpin/actors/aws/s3.py

```diff
@@ -0,0 +1,216 @@
+                    (self.option('name'), self.option('region')))
+            if e.status == 404:
+                self.log.debug('No bucket %s found' % self.option('name'))
+                raise gen.Return(None)
+
+            raise exceptions.RecoverableActorFailure(
+                'An unexpected error occurred: %s' % e)
+
+        self.log.debug('Found bucket %s' % bucket)
+        raise gen.Return(bucket)
+
+    @gen.coroutine
+    def _ensure_bucket(self):
+        """Ensures a bucket exists or does not."""
+        # Determine if the bucket already exists or not
+        state = self.option('state')
+        bucket = yield self._get_bucket()
+
+        if state == 'absent' and bucket is None:
+            self.log.debug('Bucket does not exist')
+        elif state == 'absent' and bucket:
+            yield self._delete_bucket(bucket)
+            bucket = None
+        elif state == 'present' and bucket is None:
+            bucket = yield self._create_bucket()
+        elif state == 'present' and bucket:
+            self.log.debug('Bucket exists')
+
+        raise gen.Return(bucket)
+
+    @gen.coroutine
+    def _create_bucket(self):
+        """Creates an S3 bucket if its missing.
+
+        returns:
+            <A boto.s3.Bucket object>
+        """
+        if self._dry:
+            self.log.warning('Would have created s3://%s' %
+                             self.option('name'))
+            raise gen.Return()
+
+        # This throws no exceptions, even if the bucket exists, that we know
+        # about or can expect.
+        self.log.info('Creating bucket')
+        bucket = yield self.thread(self.s3_conn.create_bucket,
+                                   self.option('name'))
+        raise gen.Return(bucket)
+
+    @gen.coroutine
+    def _delete_bucket(self, bucket):
+        """Tries to delete an S3 bucket.
+
+        args:
+            bucket: The S3 bucket object as returned by Boto
+        """
+        # Find out if there are any files in the bucket before we go to delete
+        # it. We cannot delete a bucket with files in it -- nor do we want to.
+        keys = yield self.thread(bucket.get_all_keys)
+        if len(keys) > 0:
+            raise exceptions.RecoverableActorFailure(
+                'Cannot delete bucket with keys: %s files found' % len(keys))
+
+        if self._dry:
+            self.log.warning('Would have deleted bucket %s' % bucket)
+            raise gen.Return()
+
+        try:
+            self.log.info('Deleting bucket %s' % bucket)
+            yield self.thread(bucket.delete)
+        except S3ResponseError as e:
+            if e.status == 409:
+                raise exceptions.RecoverableActorFailure(
+                    'Cannot delete bucket: %s' % e.message)
+
+    @gen.coroutine
+    def _execute(self):
+        """Executes an actor and yields the results when its finished.
+
+        raises: gen.Return(True)
+        """
+        yield self._ensure_bucket()
```

Comment:
If we're deleting a deleted bucket this actor does not produce any logs at all. Intentional?